src/uct/sm/knem/knem_md.c

```diff
@@ -176,30 +227,119 @@ static ucs_status_t uct_knem_rkey_release(uct_md_component_t *mdc, uct_rkey_t rk
+    .is_mem_type_owned = (void *)ucs_empty_function_return_zero,
+};
+
+
+static ucs_status_t uct_knem_rcache_mem_reg_cb(void *context, ucs_rcache_t *rcache,
+                                               void *arg, ucs_rcache_region_t *rregion)
+{
+    uct_knem_rcache_region_t *region = ucs_derived_of(rregion, uct_knem_rcache_region_t);
+    uct_knem_md_t *md                = context;
+    int *flags = arg;
+
+    return uct_knem_mem_reg_internal(&md->super, (void*)region->super.super.start,
+                                     region->super.super.end - region->super.super.start,
+                                     *flags, &region->key);
+}
+
+static void uct_knem_rcache_mem_dereg_cb(void *context, ucs_rcache_t *rcache,
+                                         ucs_rcache_region_t *rregion)
+{
+    uct_knem_rcache_region_t *region = ucs_derived_of(rregion, uct_knem_rcache_region_t);
+    uct_knem_md_t            *md     = context;
+
+    uct_knem_mem_dereg_internal(&md->super, &region->key);
+}
+
+static void uct_knem_rcache_dump_region_cb(void *context, ucs_rcache_t *rcache,
+                                           ucs_rcache_region_t *rregion, char *buf,
+                                           size_t max)
+{
+    uct_knem_rcache_region_t *region = ucs_derived_of(rregion, uct_knem_rcache_region_t);
+    uct_knem_key_t *key = &region->key;
+
+    snprintf(buf, max, "cookie %"PRIu64" addr %p", key->cookie, (void*)key->address);
+}
+
+static ucs_rcache_ops_t uct_knem_rcache_ops = {
+    .mem_reg     = uct_knem_rcache_mem_reg_cb,
+    .mem_dereg   = uct_knem_rcache_mem_dereg_cb,
+    .dump_region = uct_knem_rcache_dump_region_cb
+};
+
+static ucs_status_t uct_knem_md_open(const char *md_name,
+                                     const uct_md_config_t *uct_md_config,
                                      uct_md_h *md_p)
 {
+    const uct_knem_md_config_t *md_config = ucs_derived_of(uct_md_config, uct_knem_md_config_t);
     uct_knem_md_t *knem_md;
-
-    static uct_md_ops_t md_ops = {
-        .close        = uct_knem_md_close,
-        .query        = uct_knem_md_query,
-        .mem_alloc    = (void*)ucs_empty_function_return_success,
-        .mem_free     = (void*)ucs_empty_function_return_success,
-        .mkey_pack    = uct_knem_rkey_pack,
-        .mem_reg      = uct_knem_mem_reg,
-        .mem_dereg    = uct_knem_mem_dereg,
-        .is_mem_type_owned = (void *)ucs_empty_function_return_zero,
-    };
+    ucs_rcache_params_t rcache_params;
+    ucs_status_t status;
 
     knem_md = ucs_malloc(sizeof(uct_knem_md_t), "uct_knem_md_t");
     if (NULL == knem_md) {
         ucs_error("Failed to allocate memory for uct_knem_md_t");
         return UCS_ERR_NO_MEMORY;
     }
 
-    knem_md->super.ops = &md_ops;
-    knem_md->super.component = &uct_knem_md_component;
+    knem_md->super.ops         = &md_ops;
+    knem_md->super.component   = &uct_knem_md_component;
+    knem_md->reg_cost.overhead = 1200.0e-9;
```

Comment:
need to also update cma / knem iface BW to reflect the fact that KNEM is faster (the numbers should be based on actual performance data for large (~1mb) messages)

Reply:
it was original value used prior to PR

Reply:
indeed, but now we know it is wrong value, since you said there is ~15% diff

Reply:
btw, 15% is comparing with CMA, not KNEM (comparing KNEM without rcache it is about 2x difference)

Reply:
right, so the 15% diff should be also in the iface estimated bandwidth, as reported by uct_iface_query(). currently both cma and knem return 12 MB/s

Reply:
and one more note: this value is used when rcache is not enabled (failed to initialize or disabled manually)

Reply:
sorry about confusion - my comment is not directly related to that particular code line